include/zephyr/drivers/video.h

```diff
@@ -118,6 +118,66 @@ struct video_buffer {
 	uint32_t timestamp;
 };
 
+/**
+ * @brief video_frmival_type enum
+ *
+ * Supported frame interval type of a video device.
+ */
+enum video_frmival_type {
+	VIDEO_FRMIVAL_TYPE_DISCRETE = 1,
+	VIDEO_FRMIVAL_TYPE_STEPWISE = 2,
+};
+
+/**
+ * @struct video_frmival
+ * @brief Video frame interval structure
+ *
+ * Used to describe a video frame interval.
+ */
+struct video_frmival {
+	uint32_t numerator;
+	uint32_t denominator;
+};
+
+/**
+ * @struct video_frmival_stepwise
+ * @brief Video frame interval stepwise structure
+ *
+ * Used to describe the video frame interval stepwise type.
+ */
+struct video_frmival_stepwise {
+	/** minimum frame interval in seconds. */
+	struct video_frmival min;
+	/** maximum frame interval in seconds. */
+	struct video_frmival max;
+	/** frame interval step size in seconds. */
+	struct video_frmival step;
+};
+
+/**
+ * @struct video_frmival_enum
+ * @brief Video frame interval enumeration structure
+ *
+ * Used to describe the supported video frame intervals of a given format.
+ */
+struct video_frmival_enum {
+	/** frame interval index during enumeration. */
+	uint32_t index;
+	/** FourCC pixel format value (\ref video_pixel_formats). */
+	uint32_t pixelformat;
+	/** frame width in pixels. */
+	uint32_t width;
+	/** frame height in pixels. */
+	uint32_t height;
```

Comment:
I think I missed something: this struct seems redundant with [`struct video_format`](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/drivers/video.h#L43-L58).

It seems like the extra frame interval types could be an extension to the existing `struct video_format` instead?

Then there would [no need to introduce an extra API for this](https://github.com/zephyrproject-rtos/zephyr/assets/55351021/97c7e01c-7084-460d-9073-e247410e92f8) (from [here](https://static.linaro.org/connect/san19/presentations/san19-503.pdf#page=10))

In hope I am not mixing unrelated things...

Thank you very much for adding this essential point! Twice the FPS requires twice the hardware bandwidth, usually it is only possible at lower resolution.

Reply:
@josuah : Thanks. Fframe rate depends on pixel format, width and height.
Unlike other settings such as contrast, awb, etc. which are simply controlled by only the camera sensor and so, could be set which a video_cid setting, frame rate (and also format) dedicate an API because it depends on other hardware along the camera pipeline and need a runtime accordance between them.

Reply:
> I also wonder: if using set_frmival(YUYV 640x480px at 60FPS) and then set_format(RGB8 1280x720px), the width and height could be left inconsistent with the frame rate/interval for some sensors? Or the driver would have to automatically pick the right FPS.
> 
> Or would the driver have to auto-select the frame rate so that the requested size can be achieved?

- set_frmival() does not set the format (e.g. width and height), it just set the fps (but you need to provide width, height, pixel format as input) so there will be no "inconsistent" with set_fmt().
- It's the driver's responsability to handle cases that the user's required fps not supported. It can pick the nearest supported fps for that format or simply return error code.
- set_frmival() should not replace set_format() as already, the two APIs do different things, moreover set_frmival() is optional while set_format() is mandatory.

Reply:
> set_frmival() does not set the format

Ah, this is what I missed! Setting the frame interval requires width and height to be set, but they are only hints, do not change the values.

I also missed that these are all V4L2 semantics here.
These are documented in the Linux kernel.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/media/v4l2-subdev.h#n784
https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/vidioc-subdev-g-frame-interval.html

> Drivers must not return an error solely because the requested interval doesn’t match the device capabilities. They must instead modify the interval to match what the hardware can provide.
> 
> Changing the frame interval shall never change the format. Changing the format, on the other hand, may change the frame interval.

That might need links on the doc, but that is another pull request maybe.

Everything is clear now. Thank you a lot for taking the time of explaining earlier!

Reply:
> Drivers must not return an error solely because the requested interval doesn’t match the device capabilities. They must instead modify the interval to match what the hardware can provide.
> 
> Changing the frame interval shall never change the format. Changing the format, on the other hand, may change the frame interval.

Yes, I will add these in the code comments to be used as doc. Thanks !